### ISSUE TITLE
Help updating OpenBSD port to neovim 0.4.2

### ISSUE THREAD

**Issue by kelp** [neovim/neovim#11320]
Hi there,

I'm trying to update the OpenBSD port to Neovim 0.4.2 and running into some challenges with the new libluv dependency. There is no current OpenBSD port for luv, and Neovim is the only thing that requires it. I did create a luv port, but rather than use that we're hoping we can just use the bundled dependencies only for luv. But use external dependencies for everything else.

You can see the discussion here: https://marc.info/?t=156978843400002&r=1&w=2

We had hoped to do something like calling cmake with:
-DUSE_BUNDLED_GPERF=Off \
-DUSE_BUNDLED_UNIBILIUM=Off \
-DUSE_BUNDLED_LIBTERMKEY=Off \
-DUSE_BUNDLED_LIBVTERM=Off \
-DUSE_BUNDLED_LIBUV=Off \
-DUSE_BUNDLED_MSGPACK=Off \
-DUSE_BUNDLED_LUAJIT=Off \
-DUSE_BUNDLED_LUAROCKS=Off \
-DUSE_BUNDLED_LUV=On

But that doesn't seem to build ANY of the bundled deps, and the configure step fails, since it's unable to find luv. If I'm understanding the build system correctly, it will only try to build third-party deps if -DBUNDLED_DEPS=ON. Is that correct?

I've tried hacking around this by doing a separate step to first setup the third-party deps.

        mkdir -p ${WRKBUILD}/deps
        cd ${WRKBUILD}/deps
        cmake -G 'Ninja' ${CONFIGURE_ARGS} ${WRKSRC}/third-party
        ninja

But when doing the final build I get the failure you see here:

https://marc.info/?l=openbsd-ports&m=157245630805788&w=2

I'm wondering if you have a recommendation for a better way to approach this?

I'd love to get a recent neovim release into OpenBSD, and I think handling this luv dependency is the main blocker for future updates too.

Thanks for any help.

**Comment by justinmk** [comment]
Thank you very much for working on this! I wonder if @vext01 would have any advice.

**Comment by kelp** [comment]
Hey @justinmk he actually suggested I check in here. :)

**Comment by justinmk** [comment]
> I did create a luv port, but rather than use that we're hoping we can just use the bundled dependencies only for luv.

What's wrong with having a luv port?

**Comment by kelp** [comment]
> > I did create a luv port, but rather than use that we're hoping we can just use the bundled dependencies only for luv.
> 
> What's wrong with having a luv port?

See Stewart's comments here: https://marc.info/?l=openbsd-ports&m=157166003611183&w=2

He pointed out a number of issues with my luv port, and we were thinking it might be easier to just use the bundled dependencies.

Turns out maybe not so easy? I've been kind of wondering if the Neovim team had intended for anyone to make partial use of the bundled dependencies. Or is it intended to be all or none?

**Comment by justinmk** [comment]
> we were thinking it might be easier to just use the bundled dependencies.
> Turns out maybe not so easy?

Depends. Bundled deps are easy to use when building from source, since that's their purpose. For package systems, usually downloading stuff from the net is disallowed. 

> is it intended to be all or none

No, we support `USE_BUNDLED_*` so that people can use system packages for selected deps.

I haven't read any of your references yet, don't have time. If it's asking for BSD-specific I won't be able to help. If it's a question about our build-system features please post those questions here.

**Comment by justinmk** [comment]
> If I'm understanding the build system correctly, it will only try to build third-party deps if `-DBUNDLED_DEPS=ON`. Is that correct?

Maybe? If you have `-DUSE_BUNDLED_LUV=On` then you want bundled deps, so...

**Comment by kelp** [comment]
> > If I'm understanding the build system correctly, it will only try to build third-party deps if `-DBUNDLED_DEPS=ON`. Is that correct?
> 
> Maybe? If you have `-DUSE_BUNDLED_LUV=On` then you want bundled deps, so...

Sorry I should have led with info about this.

I've tried this:

 ```
mkdir -p build
cd build
cmake -G 'Ninja' -DUSE_BUNDLED_LUV=On PREFER_LUA=ON ../
BUILDING NVIM

-- The C compiler identification is Clang 8.0.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_INSTALL_PREFIX=/usr/local
-- CMAKE_BUILD_TYPE not specified, default is 'Debug'
-- MIN_LOG_LEVEL not specified, default is 1 (INFO)
-- Performing Test HAS_ACCEPTABLE_FORTIFY
-- Performing Test HAS_ACCEPTABLE_FORTIFY - Success
-- Performing Test HAVE_EXECINFO_BACKTRACE
-- Performing Test HAVE_EXECINFO_BACKTRACE - Failed
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW - Success
-- Performing Test HAS_WIMPLICIT_FALLTHROUGH_FLAG
-- Performing Test HAS_WIMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test HAS_WVLA_FLAG
-- Performing Test HAS_WVLA_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_FLAG - Success
-- Performing Test HAS_DIAG_COLOR_FLAG
-- Performing Test HAS_DIAG_COLOR_FLAG - Success
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Looking for dlopen in dl
-- Looking for dlopen in dl - not found
-- Looking for kstat_lookup in kstat
-- Looking for kstat_lookup in kstat - not found
-- Looking for kvm_open in kvm
-- Looking for kvm_open in kvm - found
-- Looking for gethostbyname in nsl
-- Looking for gethostbyname in nsl - not found
-- Looking for perfstat_cpu in perfstat
-- Looking for perfstat_cpu in perfstat - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - not found
-- Looking for sendfile in sendfile
-- Looking for sendfile in sendfile - not found
-- Found LibUV: /usr/local/lib/libuv.so.2.1 (Required is at least version "1.28.0")
-- Found Msgpack: /usr/local/lib/libmsgpackc.so.1.0 (found suitable version "2.1.5", minimum required is "1.0.0")
CMake Error at /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find LibLUV (missing: LIBLUV_LIBRARY LIBLUV_INCLUDE_DIR)
  (Required is at least version "1.30.0")
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLibLUV.cmake:29 (find_package_handle_standard_args)
  CMakeLists.txt:375 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/kelp/src/neovim-0.4.2/build/CMakeFiles/CMakeOutput.log".
See also "/home/kelp/src/neovim-0.4.2/build/CMakeFiles/CMakeError.log".
```

Is there something else I need to pass to cmake to make it use just the one bundled dep? It looks like it's not trying to use any of them.

Also just wanted to thank you for your time on this. I really appreciate you looking at this with me.

**Comment by vext01** [comment]
@kelp Thanks for following up with this btw.

@justinmk so should `cmake -G 'Ninja' -DUSE_BUNDLED_LUV=On PREFER_LUA=ON ..` work?

**Comment by justinmk** [comment]
> so should `cmake -G 'Ninja' -DUSE_BUNDLED_LUV=On PREFER_LUA=ON ..` work?

AFAIK, yes. I never use these options, but @mhinz @jamessan @blueyed may have more insight. 
 The luv dependency is fairly new, maybe there's a wrong ordering/assumption somewhere in our cmake files. 

564d415d2082b5e was intended to propagate the effects of `PREFER_LUA=ON` to the luv build, it looks like it should work: it tries `find_package(LuaJit)` and if that fails then it sends `-DWITH_LUA_ENGINE=Lua` to the luv build.

```
if(USE_BUNDLED_LUAJIT)
  ...
elseif(USE_BUNDLED_LUA)
  ...
else()
  find_package(LuaJit)
  if(LUAJIT_FOUND)
    list(APPEND LUV_CONFIGURE_COMMAND_COMMON -DWITH_LUA_ENGINE=LuaJit)
  else()
    list(APPEND LUV_CONFIGURE_COMMAND_COMMON -DWITH_LUA_ENGINE=Lua)
  endif()
endif()
```

I would focus on `third-party/cmake/BuildLuv.cmake` and add `message(FATAL_ERROR ...)` messages to check what it's doing .

**Comment by kelp** [comment]
> I would focus on `third-party/cmake/BuildLuv.cmake` and add `message(FATAL_ERROR ...)` messages to check what it's doing .

Cool thanks, this is helpful, I'll give that a try over the weekend and post here when I have an update.

**Comment by kelp** [comment]
@justinmk I'm not seeing any place in the top level CMakeLists.txt that references any USE_BUNDLED*.

For example:
``cmake -G 'Ninja' -DPREFER_LUA=ON -DUSE_BUNDLED=ON ../``

Fails because it can't find libluv. It doesn't try to use bundled deps.

However, I do see that the top level Makefile has this:

``deps: | build/.ran-third-party-cmake
ifeq ($(call filter-true,$(USE_BUNDLED)),)
        +$(BUILD_CMD) -C $(DEPS_BUILD_DIR)
endif

ifeq ($(call filter-true,$(USE_BUNDLED)),)
$(DEPS_BUILD_DIR):
        mkdir -p "$@"
build/.ran-third-party-cmake:: $(DEPS_BUILD_DIR)
        cd $(DEPS_BUILD_DIR) && \
                $(CMAKE_PRG) -G '$(BUILD_TYPE)' $(BUNDLED_CMAKE_FLAG) $(BUNDLED_LUA_CMAKE_FLAG) \
                $(DEPS_CMAKE_FLAGS) $(THIS_DIR)/third-party
endif
```
It seems like this stuff is being handled in the Makefile and not cmake?

**Comment by justinmk** [comment]
git grep USE_BUNDLED

**Comment by kelp** [comment]
Hey there, following up here. Thanks for the pointers. I think I have what I need at this point.

I stumbled upon this bit of documentation: https://github.com/neovim/neovim/wiki/Building-Neovim#third-party-dependencies

It made me realize that I was over thinking this a bit, and was actually quite close to having it working.

It also helped confirm that I'd need to build the third-party deps as separate step, and not try to rely on the top level cmake files to do it. I've since made that work.

So thanks! I'll close out this issue.
